File: data_structures/src/KeyPatterns/Hashes/HashMap/java/SubarraySumEqualsK.java
package KeyPatterns.Hashes.HashMap.java;

import java.util.HashMap;
import java.util.Map;

public class SubarraySumEqualsK {
    public static void main(String[] args) {
        int[] nums = { 1, 1, 1 };
        int k = 2;
        System.out.println(subArraySum(nums, k));
    }

    private static int subArraySum(int[] nums, int k) {
        Map<Integer, Integer> prefixMap = new HashMap<>();
        prefixMap.put(0, 1);
        int count = 0, prefixSum = 0;

        for (int num : nums) {
            prefixSum += num;
            count += prefixMap.getOrDefault(prefixSum - k, 0);
            prefixMap.put(prefixSum, prefixMap.getOrDefault(prefixSum, 0) + 1);
        }

        return count;
    }
}

File: data_structures/src/KeyPatterns/Hashes/HashMap/typescript/subarray_sum_equals_k.ts
function subArraySum(nums: number[], k: number): number {
  const prefixMap = new Map<number, number>();
  prefixMap.set(0, 1);
  let count = 0,
    prefixSum = 0;

  for (const num of nums) {
    prefixSum += num;
    count += prefixMap.get(prefixSum - k) || 0;
    prefixMap.set(prefixSum, (prefixMap.get(prefixSum) || 0) + 1);
  }

  return count;
}
console.log(subArraySum([1, 1, 1], 2));
